sw/device/lib/dif/dif_otp_ctrl.c

```diff
@@ -3,3 +3,855 @@
+      case kDifOtpCtrlStatusCodeSecret2Error:
+        field = (bitfield_field32_t){
+            .mask = OTP_CTRL_ERR_CODE_ERR_CODE_5_MASK,
+            .index = OTP_CTRL_ERR_CODE_ERR_CODE_5_OFFSET,
+        };
+        break;
+      case kDifOtpCtrlStatusCodeLifeCycleError:
+        field = (bitfield_field32_t){
+            .mask = OTP_CTRL_ERR_CODE_ERR_CODE_6_MASK,
+            .index = OTP_CTRL_ERR_CODE_ERR_CODE_6_OFFSET,
+        };
+        break;
+      case kDifOtpCtrlStatusCodeDaiError:
+        field = (bitfield_field32_t){
+            .mask = OTP_CTRL_ERR_CODE_ERR_CODE_7_MASK,
+            .index = OTP_CTRL_ERR_CODE_ERR_CODE_7_OFFSET,
+        };
+        break;
+      case kDifOtpCtrlStatusCodeLciError:
+        field = (bitfield_field32_t){
+            .mask = OTP_CTRL_ERR_CODE_ERR_CODE_8_MASK,
+            .index = OTP_CTRL_ERR_CODE_ERR_CODE_8_OFFSET,
+        };
+        break;
+      default:
+        continue;
+    }
+
+    dif_otp_ctrl_error_t err;
+    switch (bitfield_field32_read(error_codes, field)) {
+      case OTP_CTRL_ERR_CODE_ERR_CODE_0_VALUE_NO_ERROR:
+        err = kDifOtpCtrlErrorOk;
+        break;
+      case OTP_CTRL_ERR_CODE_ERR_CODE_0_VALUE_MACRO_ERROR:
+        err = kDifOtpCtrlErrorMacroUnspecified;
+        break;
+      case OTP_CTRL_ERR_CODE_ERR_CODE_0_VALUE_MACRO_ECC_CORR_ERROR:
+        err = kDifOtpCtrlErrorMacroRecoverableRead;
+        break;
+      case OTP_CTRL_ERR_CODE_ERR_CODE_0_VALUE_MACRO_ECC_UNCORR_ERROR:
+        err = kDifOtpCtrlErrorMacroUnrecoverableRead;
+        break;
+      case OTP_CTRL_ERR_CODE_ERR_CODE_0_VALUE_MACRO_WRITE_BLANK_ERROR:
+        err = kDifOtpCtrlErrorMacroBlankCheckFailed;
+        break;
+      case OTP_CTRL_ERR_CODE_ERR_CODE_0_VALUE_ACCESS_ERROR:
+        err = kDifOtpCtrlErrorLockedAccess;
+        break;
+      case OTP_CTRL_ERR_CODE_ERR_CODE_0_VALUE_CHECK_FAIL_ERROR:
+        err = kDifOtpCtrlErrorBackgroundCheckFailed;
+        break;
+      case OTP_CTRL_ERR_CODE_ERR_CODE_0_VALUE_FSM_STATE_ERROR:
+        err = kDifOtpCtrlErrorFsmBadState;
+        break;
+      default:
+        return kDifOtpCtrlError;
+    }
+    status->causes[i] = err;
+  }
+
+  return kDifOtpCtrlOk;
+}
+
+typedef struct partition_info {
+  /**
+   * The absolute OTP address at which this partition starts.
+   */
+  uint32_t start_addr;
+  /**
+   * The length of this partition, in bytes, including the digest.
+   *
+   * If the partition has a digest, it is expected to be at address
+   * `start_addr + len - sizeof(uint64_t)`.
+   */
+  uint32_t len;
+  /**
+   * The alignment mask for this partition.
+   *
+   * A valid address for this partition must be such that
+   * `addr & align_mask == 0`.
+   */
+  uint32_t align_mask;
+
+  /**
+   * Whether this is a software-managed partition with a software-managed
+   * digest.
+   */
+  bool is_software;
+} partition_info_t;
+
+static const partition_info_t kPartitions[] = {
+    // TODO: These should be provided by the gen'd header.
```

Comment:
Can you link an issue here just so the reader can check status if they see the comment.

Reply:
The issue already exists: https://github.com/lowRISC/opentitan/issues/3904
It just needs to be linked in the comment.

Reply:
Cooool! Done.